### ISSUE TITLE
JPM undefined Error: spawn EACCES

### ISSUE THREAD

**Issue by Wenzel** [mozilla-jetpack/jpm#287]
Hi !
I'm trying to run jpm on the add-on from the [tutorial](https://developer.mozilla.org/en-US/Add-ons/SDK/Tutorials/Getting_Started_%28jpm%29)
this is my add-on source tree :

~~~
.
├── data
│   ├── icon-16.png
│   ├── icon-32.png
│   └── icon-64.png
├── index.js
├── package.json
├── README.md
└── test
    └── test-index.js

2 directories, 7 files
~~~

when i launch `jpm run`, i have the following output :

~~~
JPM undefined Starting jpm run on My Jetpack Addon
Creating XPI
JPM undefined XPI created at /tmp/@source-0.0.1.xpi (30ms)
Created XPI at /tmp/@source-0.0.1.xpi
JPM undefined Creating a new profile
JPM undefined Error: spawn EACCES
JPM undefined Error: spawn EACCES
~~~

my development environement :
-  `opensuse 13.2 x64`
- `firefox 36.0.4`

I notice that `/usr/bin/firefox` was a symlink to `/usr/lib64/firefox/firefox`
and having looked at the previous issued, I thought that this one might be linked 
[ Default Firefox path for jpm doesn't work on Ubuntu #178 ](https://github.com/mozilla/jpm/issues/178)

I tried to move it and to create a hardlink 
~~~
ln /usr/lib64/firefox/firefox /usr/bin/firefox
~~~

but it didn't work.

It it an issue related to `jpm` code ?

Thanks !

**Comment by Wenzel** [comment]
Actually  you can specify the firefox binary to use in `jpm` with [the `-b` option](https://developer.mozilla.org/en-US/Add-ons/SDK/Tools/jpm#jpm_run).

So using `jpm -b /usr/lib64/firefox/firefox run` is a working solution to this issue.

**Comment by feinstaub** [comment]
Without the -b option I also get the EACCES error (jpm --version 1.0.0)

Then I did `jpm -b /usr/lib64/firefox/firefox test` and get the following console output:
```
JPM undefined Starting jpm test on My Jetpack Addon
Creating XPI
JPM undefined XPI created at /tmp/@firefoxaddonjpm-0.0.1.xpi (48ms)
Created XPI at /tmp/@firefoxaddonjpm-0.0.1.xpi
JPM undefined Creating a new profile
```

A firefox window is opened during the process but the console output does not change. I see **no unit test output**. The addon was created via jpm init. I am working under openSUSE 13.2.

**Comment by feinstaub** [comment]
I waited more than 20 seconds for something to happen. Could this be related to https://github.com/mozilla/jpm/issues/302?

**Comment by feinstaub** [comment]
On the #jetpack IRC channel was suggested to use jpm instead of cfx.

I did
jpm -b ~/dev/share/firefox-42.0a1/firefox test -p ./_pt
and got:
```
JPM undefined Error: ENOENT, stat '/home/dev/src/firefox_addon_local_filesystem_links/_p1/chrome_debugger_profile/lock'
```

see also https://github.com/mozilla/jpm/issues/287

**Comment by feinstaub** [comment]
I retested on my openSUSE machine:

I updated jpm: sudo npm install jpm -g

jpm -b ~/dev/share/firefox-42.0a1/firefox test -p ./_pt
runs through.

jpm -b ~/dev/share/firefox-42.0a1/firefox xpi
creates successfully creates xpi file (jid1-JAzC7z53jemo5Q@jetpack-0.99.38.xpi) which does not work (but that's an some other issue)

**Comment by kumar303** [comment]
it sounds like `-b` is the solution here.

**Comment by simonbcn** [comment]
In Arch Linux I have the same issue. Firefox is installed in `/usr/lib/firefox/firefox` and there is a symlink in `/usr/bin/firefox`:
```
lrwxrwxrwx 1 root root 24 mar 19 14:29 /usr/bin/firefox -> /usr/lib/firefox/firefox
```
`/usr/bin` is a standard path for executables on Linux. Why jpm doesn't respect the Linux standard?

**Comment by kumar303** [comment]
No reason. It probably just wasn't implemented correctly. Feel free to propose a patch if you're able to help fix it.

**Comment by simonbcn** [comment]
So why this issue is closed if it isn't yet solved? :confused:

**Comment by kumar303** [comment]
whoops, you're right. I didn't read the original report close enough. Using `-b` is a workaround but it's still a bug that could be fixed.

**Comment by freaktechnik** [comment]
This is fixed with https://github.com/mozilla-jetpack/node-fx-runner/pull/7 and thus is related to #297

**Comment by wildskyf** [comment]
Is this bug been fixed?
I still got that error `JPM [error] Error: spawn EACCES` when run `jpm` without `-b`. :crying_cat_face: 
I think maybe we could add a message for users so that they could know that `-b` is necessary.

**Comment by freaktechnik** [comment]
The bug is fixed in upstream fx-runner. However that version is not published to npm and neither is jpm updated to depend on a version where it would be fixed (since there is no published version of it).

**Comment by freaktechnik** [comment]
And an update to jpm, since jpm has strict version dependencies.

**Comment by kumar303** [comment]
Sorry, last time I went to release fx-runner I hit a mac os x bug and then got side tracked. I just released 1.0.2 https://www.npmjs.com/package/fx-runner If someone could test and update jpm that would be very helpful.

**Comment by freaktechnik** [comment]
Hmm, when installed via npm link I get an ENOENT that it can't find `node_modules/.bin/mkdirp`, which only exists in the jpm installation itself, so I'm not sure if that's just an npm link specific issue. This is on Linux, of course.

**Comment by kumar303** [comment]
Maybe check that your npm is up to date? This is what `npm ls` within fx-runner tells me:
````
$ npm ls mkdirp
fx-runner@1.0.2 /Users/kumar/dev/node-fx-runner
└─┬ mocha@2.1.0
  ├─┬ jade@0.26.3
  │ └── mkdirp@0.3.0
  └── mkdirp@0.5.0
````

**Comment by freaktechnik** [comment]
I guess I didn't explain the problem here well enough: it tried to find the mkdirp in the node_modules of the current work directory (probably due to the nature of how npm link works). But since mkdirp is a dependency that would be in the jpm installation it cannot be found in the current work directory if that is not the jpm root directory.

**Comment by kumar303** [comment]
You are trying to link fx-runner to jpm? The steps would be:
````
cd /path/to/src/fx-runner
npm link
cd /path/to/src/jpm
npm link fx-runner
````

**Comment by freaktechnik** [comment]
No, I linked my local copy of jpm into the global scope.

**Comment by kumar303** [comment]
It's probably unrelated but I just published 1.0.3 to fix an outdated when module that was causing webpack errors in web-ext. https://www.npmjs.com/package/fx-runner

**Comment by freaktechnik** [comment]
Hmm, the linked version works fine in an extension without a `.bin` folder in its `node_modules` (without having to specify -b to run Firefox and my environment doesn't have <var>JPM_FIREFOX_BINARY</var> set to `$(which firefox)`). I'll try to emulate a jpm installed with npm with the updated fx-runner later and see if that resolves the `mkdirp` problem.

**Comment by freaktechnik** [comment]
This is fixed for me with 1.1.3. The issues I was having were probably due to my global npm being ancient (I forgot the global npm version would be different from my local one due to nvm).

**Comment by kumar303** [comment]
fixed in https://github.com/mozilla-jetpack/node-fx-runner/pull/7
